python/cudaq/qis/qis.py

```diff
@@ -0,0 +1,271 @@
+        to `get_num_parameters`) and quantum types (`qubit`, `qvector`, `qview`).
+        """
+        opName = cls.get_name()
+        parameters = list(args)[:cls.get_num_parameters()]
+        quantumArguments = list(args)[cls.get_num_parameters():]
+        qubitIds = processQubitIds(opName, *quantumArguments)
+        controls = qubitIds[:len(qubitIds) - 1]
+        targets = [qubitIds[-1]]
+
+        for q in quantumArguments:
+            if isinstance(q, qubit) and q.is_negated():
+                x()(q)
+
+        cudaq_runtime.applyQuantumOperation(opName, parameters, controls,
+                                            targets, False, SpinOperator())
+        for q in quantumArguments:
+            if isinstance(q, qubit) and q.is_negated():
+                x()(q)
+                q.reset_negation()
+
+    @classmethod
+    def adj(cls, *args):
+        """
+        Invoke the general adjoint version of the quantum operation. 
+        The arguments can contain float parameters (of the correct number according
+        to `get_num_parameters`) and quantum types (`qubit`, `qvector`, `qview`).
+        """
+        opName = cls.get_name()
+        parameters = list(args)[:cls.get_num_parameters()]
+        quantumArguments = list(args)[cls.get_num_parameters():]
+        qubitIds = [q for q in processQubitIds(opName, *quantumArguments)]
+
+        [
+            cudaq_runtime.applyQuantumOperation(opName,
+                                                [-1 * p
+                                                 for p in parameters], [], [q],
+                                                False, SpinOperator())
+            for q in qubitIds
+        ]
+
+
+# Define our quantum operations
+h = type('h', (quantum_operation,), {'get_name': staticmethod(lambda: 'h')})
+x = type('x', (quantum_operation,), {'get_name': staticmethod(lambda: 'x')})
+y = type('y', (quantum_operation,), {'get_name': staticmethod(lambda: 'y')})
+z = type('z', (quantum_operation,), {'get_name': staticmethod(lambda: 'z')})
+s = type('s', (quantum_operation,), {'get_name': staticmethod(lambda: 's')})
+t = type('t', (quantum_operation,), {'get_name': staticmethod(lambda: 't')})
+
+rx = type(
+    'rx', (quantum_operation,), {
+        'get_name': staticmethod(lambda: 'rx'),
+        'get_num_parameters': staticmethod(lambda: 1)
+    })
+ry = type(
+    'ry', (quantum_operation,), {
+        'get_name': staticmethod(lambda: 'ry'),
+        'get_num_parameters': staticmethod(lambda: 1)
+    })
+rz = type(
+    'rz', (quantum_operation,), {
+        'get_name': staticmethod(lambda: 'rz'),
+        'get_num_parameters': staticmethod(lambda: 1)
+    })
+r1 = type(
+    'r1', (quantum_operation,), {
+        'get_name': staticmethod(lambda: 'r1'),
+        'get_num_parameters': staticmethod(lambda: 1)
+    })
+
+ch = type('ch', (quantum_operation,), {'get_name': staticmethod(lambda: 'ch')})
```

Comment:
I don't see any support for `<op>dg` (e.g. tdg) here, which is what is used for Adjoint(T) in the kernel builder. It would be nice to add that as well.

Reply:
In the new language, the equivalent is `t.adj(...)`, we can add convenience ops though to provide `tdg(...)` like we did for `x.ctrl(...) -> cx(...)`

Reply:
Yes, that's exactly my thinking - it would be good to make it as straightforward as possible to migrate code, and having a 1:1 equivalent here hence seems like a good thing to add.